README.md
libmaus2
========

libmaus2 is a collection of data structures and algorithms. It contains

 - I/O classes (single byte and UTF-8)
 - bitio classes (input, output and various forms of bit level manipulation)
 - text indexing classes (suffix and LCP array, fulltext and minute (FM), ...)
 - BAM sequence alignment files input/output (simple and collating)

and many lower level support classes.

-------------------------------------------------------------------------------

The main development branch of libmaus2 is hosted by github at

https://github.com/gt1/libmaus2

Release packages can be found at

https://github.com/gt1/libmaus2/releases

Please make sure to choose a package containing the word "release" in it's name if you
intend to compile libmaus2 for production (i.e. non development) use.

-------------------------------------------------------------------------------

Compilation
-----------

libmaus2 uses the GNU autoconf/automake tool set. It can be compiled on Linux
using:

	libtoolize
	aclocal
	autoreconf -i -f
	./configure
	make

Running autoreconf requires a complete set of tools including autoconf, automake,
autoheader, aclocal and libtool.

The release packages come with a configure script, so running libtoolize, aclocal etc should not be necessary, i.e.

	./configure
	make

should be sufficient to compile libmaus2.

A full list of configuration parameters can be obtained by calling

	./configure --help

libmaus2 can use functionality from several other code bases. This includes:

 - snappy [http://google.github.io/snappy/] : a fast and lightweight Lempel-Ziv
   type compression/decompression library. This can be used for the compression
   of temporary files (as used in name collating BAM input for instance).
 - io_lib [http://sourceforge.net/p/staden/code/HEAD/tree/io_lib/] : This is
   part of the Staden package. libmaus2 can use this library for SAM and CRAM
   file input.

The compilation on Darwin (MacOS X) may require the installation of the 
following packages:

 - pkg-config [http://www.freedesktop.org/wiki/Software/pkg-config/]
 - boost [http://www.boost.org/]

pkg-config will be installed on most Linux systems. The boost libraries
are required if libmaus2 is to be compiled using non recent versions of the
GNU compiler, as they contain some classes offering functionality that
has meanwhile been added to the C++ standard (shared_ptr, unique_ptr,
unordered_map, ...).

Compilation of boost for libmaus2 on Darwin
------------------------------------------

The following lines called in an unpacked boost source tree should be 
sufficient to compile and install a version of the boost libraries usable
by libmaus2:

	INSTPREFIX=${HOME}/libs/boost
	./bootstrap.sh --prefix=${INSTPREFIX}
	./b2 --prefix=${INSTPREFIX} --build-type=minimal \
		--layout=system toolset=darwin variant=release link=static \
		address-model=32_64 threading=multi cxxflags="-arch i386 -arch x86_64"
	./b2 --prefix=${INSTPREFIX} --build-type=minimal \
		--layout=system toolset=darwin variant=release link=static \
		address-model=32_64 threading=multi cxxflags="-arch i386 -arch x86_64" install

Compilation of libmaus2 on Darwin
--------------------------------

After installing boost, libmaus2 can be compiled and installed in ${HOME}/libmaus2 using:

	export CPPFLAGS="-I${HOME}/libs/boost/include/"
	export CFLAGS="-arch i386 -arch x86_64 -O2"
	export CXXFLAGS="-arch i386 -arch x86_64 -O2"
	export LDFLAGS="-L${HOME}/libs/boost/lib/"
	export LIBS="-arch i386 -arch x86_64"
	export CC="/usr/bin/gcc"
	export CXX="/usr/bin/g++"
	export CPP="/usr/bin/cpp"
	export CXXCPP="/usr/bin/cpp"
	bash configure --prefix=${HOME}/libmaus2 --disable-shared-libmaus2 --disable-asm \
		--disable-dependency-tracking
	make install


ChangeLog
libmaus2 (2.0.499-1) unstable; urgency=medium

  * add access for true/haplo flag in OverlapData/Interface

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 06 Jun 2018 12:11:14 +0200

libmaus2 (2.0.498-1) unstable; urgency=medium

  * add haplo flag for dazzler alignments

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jun 2018 17:02:44 +0200

libmaus2 (2.0.497-1) unstable; urgency=medium

  * fix warning

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jun 2018 15:23:54 +0200

libmaus2 (2.0.496-1) unstable; urgency=medium

  * fix warning

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 04 Jun 2018 16:52:01 +0200

libmaus2 (2.0.495-1) unstable; urgency=medium

  * fix min call by cast

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 25 May 2018 15:23:32 +0200

libmaus2 (2.0.494-1) unstable; urgency=medium

  * add missing include in ArgParser.cpp

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 25 May 2018 14:41:21 +0200

libmaus2 (2.0.493-1) unstable; urgency=medium

  * add getTracePoints function OverlapData

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 23 May 2018 18:28:25 +0200

libmaus2 (2.0.492-1) unstable; urgency=medium

  * add pointer types in FastaPeeker

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 23 May 2018 09:45:21 +0200

libmaus2 (2.0.491-1) unstable; urgency=medium

  * add putARead and putBRead in OverlapData

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 18 May 2018 10:13:34 +0200

libmaus2 (2.0.490-1) unstable; urgency=medium

  * update debian control files

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 17 May 2018 10:34:55 +0200

libmaus2 (2.0.489-1) unstable; urgency=medium

  * add typedef in SortingBufferedOutputFile

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 15 May 2018 20:17:36 +0200

libmaus2 (2.0.488-1) unstable; urgency=medium

  * add OverlapAdapter class

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 15 May 2018 12:10:21 +0200

libmaus2 (2.0.487-1) unstable; urgency=medium

  * add missing pass through of comparator in MergingReadBack

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 14 May 2018 10:24:36 +0200

libmaus2 (2.0.486-1) unstable; urgency=medium

  * add sortthreads option in reduce/sort methods in SortingBufferedOutputFile

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 09 May 2018 13:58:39 +0200

libmaus2 (2.0.485-1) unstable; urgency=medium

  * add missing initialisation of slotfilled field in GenericPeeker

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 09 May 2018 11:11:52 +0200

libmaus2 (2.0.484-1) unstable; urgency=medium

  * avoid endless loop in thresLow of Poisson

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 08 May 2018 12:16:26 +0200

libmaus2 (2.0.483-1) unstable; urgency=medium

  * add getOffset and checkSize in ContainerDescriptionList

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 07 May 2018 17:31:30 +0200

libmaus2 (2.0.482-1) unstable; urgency=medium

  * add output of modcall flag in Command

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 07 May 2018 09:54:25 +0200

libmaus2 (2.0.481-1) unstable; urgency=medium

  * add getAbsProgName in ArgParser
  * add getCurDir in PathTools
  * update getAbsProgName in ArgInfo

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 06 May 2018 10:06:55 +0200

libmaus2 (2.0.480-1) unstable; urgency=medium

  * add modcall field in Command class

 -- German Tischler <tischler@mpi-cbg.de>  Sat, 05 May 2018 22:23:06 +0200

libmaus2 (2.0.479-1) unstable; urgency=medium

  * add Base64 class

 -- German Tischler <tischler@mpi-cbg.de>  Sat, 05 May 2018 20:25:36 +0200

libmaus2 (2.0.478-1) unstable; urgency=medium

  * avoid unnecessary clearing of buffer in BufferedOutput

 -- German Tischler <tischler@mpi-cbg.de>  Sat, 05 May 2018 00:43:44 +0200

libmaus2 (2.0.477-1) unstable; urgency=medium

  * use absolute paths for symlink sources in DatabaseFile
  * add PathTools

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 04 May 2018 14:36:37 +0200

libmaus2 (2.0.476-1) unstable; urgency=medium

  * fix serialise in ContainerDescriptionList

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 02 May 2018 15:43:25 +0200

libmaus2 (2.0.475-1) unstable; urgency=medium

  * add LIBMAUS2_USE_ASSEMBLY in AlignmentOneAgainstManyFactory check

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 02 May 2018 12:39:40 +0200

libmaus2 (2.0.474-1) unstable; urgency=medium

  * add replace method in ContainerDescriptionList
  * add serialisedSize method in ContainerDescription
  * only try to use I386CacheLineSize if assembly is enabled

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 02 May 2018 10:42:14 +0200

libmaus2 (2.0.473-1) unstable; urgency=medium

  * rewrite Command dispatching

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 30 Apr 2018 10:20:37 +0200

libmaus2 (2.0.472-1) unstable; urgency=medium

  * fix splitting in splitDbRLTwo

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 27 Apr 2018 17:10:44 +0200

libmaus2 (2.0.471-1) unstable; urgency=medium

  * add writeSplitDatabase in DatabaseFile

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 26 Apr 2018 17:30:20 +0200

libmaus2 (2.0.470-1) unstable; urgency=medium

  * add splitDbRLTwo in DatabaseFile

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 26 Apr 2018 13:45:10 +0200

libmaus2 (2.0.469-1) unstable; urgency=medium

  * add libmaus2/util/FileEnumerator.hpp in Makefile.am
  * add enumerateTracks method in DatabaseFile class
  * add splitDbRLPrefixStrict in DatabaseFile

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 26 Apr 2018 11:32:38 +0200

libmaus2 (2.0.468-1) unstable; urgency=medium

  * fix NPLNoTrace for case one input string is empty
  * add serialise method in DatabaseFile

 -- German Tischler <tischler@mpi-cbg.de>  Wed, 25 Apr 2018 13:50:50 +0200

libmaus2 (2.0.467-1) unstable; urgency=medium

  * add testnnplocalaligner

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 13 Apr 2018 16:53:21 +0200

libmaus2 (2.0.466-1) unstable; urgency=medium

  * add prepend method in AlignmentTraceContainer


src/libmaus/util/md5.h
/*
  Copyright (C) 1999, 2002 Aladdin Enterprises.  All rights reserved.

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.

  L. Peter Deutsch
  ghost@aladdin.com

 */
/* $Id: md5.h,v 1.1 2012-05-28 22:35:59 gt1 Exp $ */
/*
  Independent implementation of MD5 (RFC 1321).

  This code implements the MD5 Algorithm defined in RFC 1321, whose
  text is available at
	http://www.ietf.org/rfc/rfc1321.txt
  The code is derived from the text of the RFC, including the test suite
  (section A.5) but excluding the rest of Appendix A.  It does not include
  any code or documentation that is identified in the RFC as being
  copyrighted.

  The original and principal author of md5.h is L. Peter Deutsch
  <ghost@aladdin.com>.  Other authors are noted in the change history
  that follows (in reverse chronological order):

  2002-04-13 lpd Removed support for non-ANSI compilers; removed
	references to Ghostscript; clarified derivation from RFC 1321;
	now handles byte order either statically or dynamically.
  1999-11-04 lpd Edited comments slightly for automatic TOC extraction.
  1999-10-18 lpd Fixed typo in header comment (ansi2knr rather than md5);
	added conditionalization for C++ compilation from Martin
	Purschke <purschke@bnl.gov>.
  1999-05-03 lpd Original version.
 */

#ifndef md5_INCLUDED
#  define md5_INCLUDED

/*
 * This package supports both compile-time and run-time determination of CPU
 * byte order.  If ARCH_IS_BIG_ENDIAN is defined as 0, the code will be
 * compiled to run only on little-endian CPUs; if ARCH_IS_BIG_ENDIAN is
 * defined as non-zero, the code will be compiled to run only on big-endian
 * CPUs; if ARCH_IS_BIG_ENDIAN is not defined, the code will be compiled to
 * run on either big- or little-endian CPUs, but will run slightly less
 * efficiently on either one than if ARCH_IS_BIG_ENDIAN is defined.
 */

typedef unsigned char md5_byte_t; /* 8-bit byte */
typedef unsigned int md5_word_t; /* 32-bit word */

/* Define the state of the MD5 Algorithm. */
typedef struct md5_state_s {
    md5_word_t count[2];	/* message length in bits, lsw first */
    md5_word_t abcd[4];		/* digest buffer */
    md5_byte_t buf[64];		/* accumulate block */
} md5_state_t;

#ifdef __cplusplus
extern "C" 
{
#endif

/* Initialize the algorithm. */
void md5_init(md5_state_t *pms);

/* Append a string to the message. */
void md5_append(md5_state_t *pms, const md5_byte_t *data, int nbytes);

/* Finish the message and return the digest. */
void md5_finish(md5_state_t *pms, md5_byte_t digest[16]);

#ifdef __cplusplus
}  /* end extern "C" */
#endif

#endif /* md5_INCLUDED */
